File: algorithms/code/leetcode/lc219_contains_duplicate_ii/lc219_contains_duplicate_ii.py
# https://leetcode.com/problems/contains-duplicate-ii/

import collections
import unittest


class Solution:
    def containsNearbyDuplicate(self, nums: list[int], k: int) -> bool:
        collected: set[int] = set()
        ordered: collections.deque = collections.deque()
        for num in nums:
            if num in collected:
                return True
            ordered.append(num)
            collected.add(num)
            if len(ordered) > k:
                collected.remove(ordered.popleft())
        return False


class TestCode(unittest.TestCase):
    def test_0_5_7__2(self) -> None:
        self.assertFalse(Solution().containsNearbyDuplicate([0, 5, 7], 2))

    def test_0_5_7_5__2(self) -> None:
        self.assertTrue(Solution().containsNearbyDuplicate([0, 5, 7, 5], 2))

    def test_0_5_7_10_5__2(self) -> None:
        self.assertFalse(
            Solution().containsNearbyDuplicate([0, 5, 7, 10, 5], 2)
        )

File: algorithms/code/leetcode/lc219_contains_duplicate_ii/lc219_contains_duplicate_ii.rs
// https://leetcode.com/problems/contains-duplicate-ii/
// #easy

use std::collections::{HashSet, VecDeque};

pub struct Solution;

impl Solution {
    pub fn contains_nearby_duplicate(nums: &[i32], k: usize) -> bool {
        let mut set = HashSet::new();
        let mut ordered = VecDeque::new();
        for num in nums {
            if set.contains(num) {
                return true;
            }
            ordered.push_back(num);
            set.insert(num);
            if ordered.len() > k {
                set.remove(ordered.pop_front().unwrap());
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_57_and_2() {
        assert!(!Solution::contains_nearby_duplicate(&[0, 5, 7], 2));
    }

    #[test]
    fn test_575_and_2() {
        assert!(Solution::contains_nearby_duplicate(&[0, 5, 7, 5], 2));
    }

    #[test]
    fn test_5705_and_2() {
        assert!(!Solution::contains_nearby_duplicate(&[0, 5, 7, 10, 5], 2));
    }
}
